File: python/zebra-puzzle/zebra_puzzle.py
from itertools import permutations

nationalities = ['Englishman', 'Spaniard', 'Ukrainian', 'Norwegian', 'Japanese']
colors = ['red', 'green', 'yellow', 'blue', 'ivory']
pets = ['dog', 'snails', 'fox', 'horse', 'zebra']
drinks = ['coffee', 'tea', 'water', 'milk', 'orange juice']
smokes = ['Old Gold', 'Kools', 'Chesterfields', 'Lucky Strike', 'Parliaments']

for n in permutations(nationalities):
    if n[0] != 'Norwegian': continue #10
    for c in permutations(colors):
        if c.index('green') != c.index('ivory') + 1: continue #6
        if abs(n.index('Norwegian') - c.index('blue')) != 1: continue #15
        if n.index('Englishman') != c.index('red'): continue #2
        for p in permutations(pets):
            if n.index('Spaniard') != p.index('dog'): continue #3
            for d in permutations(drinks):
                if d.index('milk') != 2: continue #9
                if d.index('coffee') != c.index('green'): continue #4
                if d.index('tea') != n.index('Ukrainian'): continue #5
                for s in permutations(smokes):
                    if s.index('Old Gold') != p.index('snails'): continue #7
                    if s.index('Kools') != c.index('yellow'): continue #8
                    if abs(s.index('Chesterfields') - p.index('fox')) != 1: continue #11
                    if abs(s.index('Kools') - p.index('horse')) != 1: continue #12
                    if s.index('Lucky Strike') != d.index('orange juice'): continue #13
                    if s.index('Parliaments') != n.index('Japanese'): continue #14
                    nationalities, colors, pets, drinks, smokes = n, c, p, d, s

def drinks_water():
    return nationalities[drinks.index('water')]


def owns_zebra():
    return nationalities[pets.index('zebra')]

File: ruby/zebra-puzzle/zebra_puzzle.rb
class ZebraPuzzle
  NATIONALITIES = %w[Englishman Norwegian Japanese Ukrainian Spaniard]
  COLORS        = %w[red green yellow ivory blue]
  PETS          = %w[dog snail fox horse zebra]
  ACTIVITY      = %w[dancing painting reading chess football]
  DRINKS        = %w[coffee tea milk oj water]

  def self.groups(type)
    NATIONALITIES.permutation.each { |n|
      next if n[0] != "Norwegian"
      COLORS.permutation.each { |c|
        next if c.index("red") != n.index("Englishman")
        next if c.index("green") != c.index("ivory") + 1
        next if (n.index("Norwegian") - c.index("blue")).abs != 1
        PETS.permutation.each { |p|
          next if p.index("dog") != n.index("Spaniard")
          DRINKS.permutation.each { |d|
            next if d.index("milk") != 2
            next if c.index("green") != d.index("coffee")
            next if n.index("Ukrainian") != d.index("tea")
            ACTIVITY.permutation.each { |a|
              next if p.index("snail") != a.index("dancing")
              next if c.index("yellow") != a.index("painting")
              next if (p.index("fox") - a.index("reading")).abs != 1
              next if (p.index("horse") - a.index("painting")).abs != 1
              next if d.index("oj") != a.index("football")
              next if n.index("Japanese") != a.index("chess")
              return type == :drinks ? d.zip(n).to_h : p.zip(n).to_h
            }
          }
        }
      }
    }
  end
  def self.water_drinker
    ZebraPuzzle.groups(:drinks)["water"]
  end
  def self.zebra_owner
    ZebraPuzzle.groups(:pets)["zebra"]
  end
end
